prow/config/cache.go

```diff
@@ -357,6 +367,16 @@ func (cache *InRepoConfigCache) get(
 		"cache_hit":         cacheHit,
 	}).Debug("Duration for resolving inrepoconfig cache.")
 
+	// Record the number of items in the cache.
+	org, repo, _ := keyToOrgRepo(key)
+	cacheUsageSize := cache.Len()
+	inRepoConfigCacheMetrics.cacheUsageSize.WithLabelValues(org, repo).Set(float64(cacheUsageSize))
```

Comment:
This gauge will not be accurate since this only updates the gauge when we create or fetch a cache entry for the repo. Evictions are not detected until the next time a cache entry is fetched or created for the repo which could be a long delay, especially for less active repos.

Reply:
Ah, true. I guess I can just have a separate goroutine that checks periodically instead (maybe every minute?).

Reply:
FWIW that still won't properly detect evictions when the count for a repo reaches 0. We could mitigate this by keeping track of the repos we've seen since startup and always serving an entry for those (observing a value of 0 if they don't appear in the cache).

> Actually I guess doing it every Sync loop is probably easiest

I think a separate goroutine is more appropriate. It won't be blocked by the sync loop and can have an independent period that is more appropriate given our scrape interval. (Specifically we should update the metrics at the same frequency of the scrape interval or faster). If we use a separate goroutine we can also encapsulate all this logic in the inrepo config cache itself rather than making each consumer of the cache (e.g. the different prow components) responsible for tracking this.

Reply:
> We could mitigate this by keeping track of the repos we've seen since startup and always serving an entry for those (observing a value of 0 if they don't appear in the cache).

Ack.

> I think a separate goroutine is more appropriate.

I was only thinking of using the Sync interval for simplicity of implementation (it would almost be a line-by-line move of the lines in this PR's diff into the Gerrit component). If we do have this metric as part of the caching code, then I guess when we create a new cache, that's the point we start reporting usage metrics.

> (Specifically we should update the metrics at the same frequency of the scrape interval or faster)

Ack.

>  If we use a separate goroutine we can also encapsulate all this logic in the inrepo config cache itself rather than making each consumer of the cache (e.g. the different prow components) responsible for tracking this.

True, but we should still have a label for the component using the cache because otherwise I assume these metrics would end up overwriting each other as they are currently used by multiple components.

Reply:
> True, but we should still have a label for the component using the cache because otherwise I assume these metrics would end up overwriting each other as they are currently used by multiple components.

We can already distinguish which component metrics came from via one of the metric labels that is automatically populated by prometheus. IIRC it is named something like `instance` or `job`